### DISCUSSION TITLE
Using the new kotlin native memory model?

### DISCUSSION THREAD

**Discussion by plusmobileapps** [badoo/Reaktive#651] [Q&A] [answered]
With the new [kotlin native memory model](https://blog.jetbrains.com/kotlin/2021/08/try-the-new-kotlin-native-memory-manager-development-preview/) being announced, is there anything that would need to be done in reaktive to be able to use the new model in a project? It does look like there [might need to be some extra configuration](https://github.com/JetBrains/kotlin/blob/master/kotlin-native/NEW_MM.md) in the project to specifically look for the new model?

**Comment by arkivanov** [comment, marked as answer]
As far as I remember, from the consumer's side the current version should be compatible with the new memory model. So it should be already possible to use Reaktive in a project with the new memory model enabled. Please let us know if something is not working as expected.

But from the Reaktive's side, we will migrate to the new memory model once it's stable-ish and widely used by the community. There are certain optimisations can be made in the library when the current memory model's limitations are lifted.  It should improve performance in various cases, simplify the code, etc.

**Reply by plusmobileapps** [reply]
Thanks, that is good to know. I will try it out in my project and let you know if I run into any problems. Will be helpful to switch over given the handful of times I've hit the InvalidMutabilityException on native.

**Reply by arkivanov** [reply]
Yeah, please give it a try! However the new memory model may be even less predictable due to its experimental status. Please share your experience!

**Reply by plusmobileapps** [reply]
Just tried the new native memory model with Kotlin 1.6.0 and was pretty simple to enable adding this into `gradle.properties`.

```
kotlin.native.binary.memoryModel=experimental
```


My project is getting more complicated but is simple enough where I don't see a performance hit yet using the new memory model and it actually helped fix a crash when navigating between configurations in decompose. Not sure what I was doing wrong, but when observables were unsubscribing there was a dereferenced exception because it was unsubscribing from a different thread 🤷  Looks promising so far though! 🚀

**Reply by arkivanov** [reply]
Glad to hear it works for you, thanks for the feedback! The fact that it crashed with the old memory model worries me. The old memory is strict, and usually crashes mean something is wrong with the code, not with the memory model. So personally I would be curious and investigated the crash.
